daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/v2/Converter.scala

```diff
@@ -33,12 +33,18 @@ object Converter extends script.ConverterMethods(StablePackagesV2) {
       ) => Either[String, TemplateChoiceSignature],
       translator: preprocessing.ValueTranslator,
       result: ScriptLedgerClient.ExerciseResult,
+      enableContractUpgrading: Boolean = false,
   ) = {
     for {
       choice <- Name.fromString(result.choice)
       c <- lookupChoice(result.templateId, result.interfaceId, choice)
       translated <- translator
-        .strictTranslateValue(c.returnType, result.result)
+        .translateValue(
+          c.returnType,
+          result.result,
+          if (enableContractUpgrading) preprocessing.ValueTranslator.Config.Upgradeable
+          else preprocessing.ValueTranslator.Config.Strict,
+        )
```

Comment:
I'm not sure how we should go about converting the resulting transaction tree. I see two related problems with this approach, one about observability and one more practical:

1. Our users would likely want to know what choice body was _actually_ used, e.g. if I create a `v1` contract, vet `v2` and then exercise choice `C` with return type `R` (regardless of which type I use on submission), I would like to see that the engine actually picked `v2:C`. 
  IIUC, if I submit the `exerciseCmd` with `v1:C`, `internalSubmit` will remember that I expect a `v1:R` (because that's the typing relation we get from `exerciseCmd`) and convert the `v2:R` result into a `v1:R`, thus losing that information and leading us to 
2. if `v2:R` has (`Optional`) fields not present in `v1:R`, and the definition of `v2:C` returns `Some` values for those fields, which appears to be allowed by the current design doc
    > for each choice `c1` with type `t1` from template `x1`:
    > - there is one corresponding choice `c2` in template `x2` with the same name;
    > - for the return type `t2` of choice `c2`, `t2` upgrades `t1`;
    
    "upgrading" that `v2:R` value to the expected `v1:R` result will fail, even if the transaction actually succeeded.

It would seem like the second problem goes away if we submit the `exerciseCmd` with `v2:C`, since the result could only be a `v1:R` or a `v2:R`, and that is true in the closed universe of `daml test`, but considering that `daml-script` is also used for interacting with real ledgers, such a script written against `v1` definitions would fail (again, the _script_ would fail, even if the transactions it submits succeed!) as soon as `v2` is uploaded and vetted - IIUC, one of the requirements for upgrades is that clients shouldn't have to update their code.

At the core of this is the typing relation imposed by `class HasExercise`

```haskell
-- | Exposes `exercise` function. Part of the `Choice` constraint.
class HasExercise t c r | t c -> r where
  -- | Exercise a choice on the contract with the given contract ID.
  exercise : ContractId t -> c -> Update r
```

"Given a template type `t` and a choice type `c`, there is exactly one result type `r` that we could get from exercising a choice of type `c` on a contract of type `t`". This is no longer true in a world with upgrades. We could restrict choice upgrades:

> for each choice `c1` with type `t1` from template `x1`:
> - there is one corresponding choice `c2` in template `x2` with the same name;
> - for the return type `t2` of choice `c2`, ~`t2` upgrades `t1`~ `t2` is structurally equivalent to `t1`;

but this would force users to add a new choice `v2:D` whenever an upgrade of `v1:C` would change the return type, in the process likely forbidding `v2:C` with `assert false`, which would similarly break existing clients.

Another tempting solution would be for `daml-script` to _unsafely_ downgrade the results coming from newer versions, but this runs the risk of breaking invariants: if the script that exercises `v1:C` then creates a contract in terms of the `v1:R` it sees, it would ignore potentially crucial information in the dropped fields.

Reply:
Perhaps what we want is a more expressive variant of `exerciseCmd`:

```haskell
exerciseCmd      : Choice t c r => ContractId t -> c -> Commands r
fancyExerciseCmd : Choice t c r => ContractId t -> c -> Commands (FancyChoiceResult r)

data FancyChoiceResult r = FancyChoiceResult LedgerValue

-- | succeeds only if the intended `r` matches* the actual type
-- *matches structurally or nominally, TBD
projectFancyChoiceResult : Choice t c r => FancyChoiceResult r -> Maybe r

-- | always succeeds but might drop important fields, hence `unsafe*`
unsafeProjectFancyChoiceResult : Choice t c r => FancyChoiceResult r -> r

-- | always succeeds. `fst . breakFancyChoiceResult === unsafeProjectFancyChoiceResult` and
-- `snd . breakFancyChoiceResult` only contains the fields not in `r`.
breakFancyChoiceResult : Choice t c r => FancyChoiceResult r -> (r, ExtraFields)

-- | Just one possible implementation
type ExtraFields = Map Text LedgerValue
```

Reply:
This is very interesting case, good spot!
I not even sure what the correct behaviour here should be, there is a possible `AnyEquivalentTemplate` type we could use, which allows coercion to equivalent templates. if we put this in the tree, then its on the user to decide what to do when their client gets something that can't be safely downgraded to the type it expects. But it'd argue that if that new data is being returned, the semantics have changed, the client can't possible foresee this.
I would think surely the first thing you update is the clients no? in such a way that they handle both versions, then you make the switch

Reply:
Ah, i missed the second message, yeah we could try something like this

Reply:
I will however say that the daml-script upgrades support is internal, and I don't need tree support right now. So I'm tempted to say that submitTree does _not_ support upgrades yet.
Naturally though, this same issue leaks into all the submit types, if we ever want to coerce down our types to something an outdated client can understand

Reply:
Having the extrafields might be nice for logging, but I really don't think an older client can do anything with these except error. Its received data that is too new, and as such, all its assumptions about _existing_ data may have changed.
Take the upgrade of a FrancsCoin to a GenericCoin that now includes the currency, for example.

Reply:
Also, would `unsafeProjectFancyChoiceResult` still fill missing fields with none? just drop the "Some" requirement for additional fields?

Reply:
> But it'd argue that if that new data is being returned, the semantics have changed, the client can't possible foresee this.
I would think surely the first thing you update is the clients no? in such a way that they handle both versions, then you make the switch

this is a fair point, I'm writing under the (months-old) assumption that "things should just:tm: work". I think saying "you should upgrade your clients first" is also a valid answer if we make that the spec.

> Ah, i missed the second message, yeah we could try something like this

asynchronous submission what? :P

> I will however say that the daml-script upgrades support is internal, and I don't need tree support right now. So I'm tempted to say that submitTree does not support upgrades yet.

this is fair too!

> Naturally though, this same issue leaks into all the submit types, if we ever want to coerce down our types to something an outdated client can understand

yup!

> Having the extrafields might be nice for logging, but I really don't think an older client can do anything with these except error. Its received data that is too new, and as such, all its assumptions about existing data may have changed.
Take the upgrade of a FrancsCoin to a GenericCoin that now includes the currency, for example.

indeed, but at least this would be explicitly checked by their script rather than silently succeeding while dropping info, or failing after having committed the transaction - though a structural `projectFancyChoiceResult` would be enough for that

Reply:
> this is a fair point, I'm writing under the (months-old) assumption that "things should just™️ work". I think saying "you should upgrade your clients first" is also a valid answer if we make that the spec.

Yeah I need to check with Bas what the expectation is here, but I really can't see anything working with the "old stuff should keep working" mentality.

> asynchronous submission what? :P

`trySubmitCommentConcurrently`

> indeed, but at least this would be explicitly checked by their script rather than silently succeeding while dropping info, or failing after having committed the transaction - though a structural `projectFancyChoiceResult` would be enough for that

Thats fair, fortunately we haven't made any promises about daml-script upgrades support, so we have time to work this out

Reply:
> Also, would `unsafeProjectFancyChoiceResult` still fill missing fields with none? 

ah right, I didn't handle this case :) I think it would be fine for it to fill missing fields with `None` (though the name "project" becomes a bit of a stretch). 

> just drop the "Some" requirement for additional fields?

I think it would have to drop any additional fields, `None` or `Some`, but I think it would still have to check that they are `Optional` (to complement the compile time check that all new fields are `Optional`).

Reply:
> but I really can't see anything working with the "old stuff should keep working" mentality.

that's two of us 😄 

> `trySubmitCommentConcurrently`

`LOCAL_VERDICT_LOCKED_COMMENT_THREAD` :(

> Thats fair, fortunately we haven't made any promises about daml-script upgrades support, so we have time to work this out

awesome 😄

Reply:
I think the basic implementation of script should mirror what we're doing in the engine for importValue and the other clients (including codegen). This means, when downgrading v2:R to v1:R, we should drop extra None fields, and error out if we encounter any additional fields with Some. 

There's a good chance we'll allow for some more general behaviour for this later on.